From 709a054b6fb0bc7d4fc5aaafc92bf4fd51cfb076 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Tue, 9 Sep 2025 11:20:00 +0800
Subject: [PATCH 232/254] mtk: hostapd: refactor link-add operation to support
 single link addition

1. leverage interface initialization flow to add a new link to an AP MLD
2. allow the new link to use the same interface name as an existing AP
   MLD

Part of this change is referred from QCA's hostapd patch:
r44-001-hostapd-ML-reconfiguration-link-addition-support.patch
(https://gitlab.com/prpl-foundation/prplos/feeds/feed-qca/-/blob/dev/ath13.1/feeds/hostapd/patches/r44-001-hostapd-ML-reconfiguration-link-addition-support.patch?ref_type=heads)

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 hostapd/ctrl_iface.c | 102 ++-----------------------------------------
 src/ap/hostapd.c     |  58 ++++++++++++++++--------
 2 files changed, 42 insertions(+), 118 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index ba95fcb81..877a57dd1 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -87,6 +87,8 @@
 static void hostapd_ctrl_iface_send(struct hostapd_data *hapd, int level,
 				    enum wpa_msg_type type,
 				    const char *buf, size_t len);
+static int hostapd_ctrl_iface_add(struct hapd_interfaces *interfaces,
+				  char *buf);
 
 
 static int hostapd_ctrl_iface_attach(struct hostapd_data *hapd,
@@ -4178,103 +4180,6 @@ static int hostapd_ctrl_iface_link_remove(struct hostapd_data *hapd, char *cmd,
 }
 
 
-static int hostapd_ctrl_iface_link_add(struct hostapd_data *hapd, char *cmd,
-				       char *buf, size_t buflen)
-{
-	struct hapd_interfaces *interfaces = hapd->iface->interfaces;
-	struct hostapd_iface *iface = NULL;
-	struct hostapd_data *h;
-	struct hostapd_config *conf;
-	const char *ifname, *conf_file, *phy;
-	u16 old_valid_links = 0;
-	bool hapd_existed = false;
-	char *pos, *tmp;
-	int i, ret = -1;
-	size_t len;
-
-	if (!hapd || !hapd->conf->mld_ap || !hapd->mld) {
-		wpa_printf(MSG_ERROR,
-			   "Trying to add link to non-MLD AP or non-existed AP");
-		return -1;
-	}
-
-	if (os_strncmp(cmd, "bss_config=", 11))
-		return -1;
-
-	len = os_strlen(cmd) + 1;
-	tmp = os_malloc(len);
-	if (!tmp)
-		return -1;
-
-	os_snprintf(tmp, len, "%s", cmd);
-	phy = tmp + 11;
-	pos = os_strchr(phy, ':');
-	if (!pos)
-		goto out;
-	*pos++ = '\0';
-	conf_file = pos;
-	if (!os_strlen(conf_file))
-		goto out;
-
-	conf = interfaces->config_read_cb(conf_file);
-	if (!conf)
-		goto out;
-
-	ifname = conf->bss[0]->iface;
-	if (ifname[0] != '\0' &&
-	    os_strncmp(ifname, hapd->conf->iface, sizeof(hapd->conf->iface))) {
-		wpa_printf(MSG_ERROR,
-			   "Interface name %s mismatch (expected %s)",
-			   ifname, hapd->conf->iface);
-		hostapd_config_free(conf);
-		goto out;
-	}
-
-	if (!conf->bss[0]->mld_ap) {
-		wpa_printf(MSG_ERROR, "The added interface is not MLD AP");
-		hostapd_config_free(conf);
-		goto out;
-	}
-
-	for (i = 0; i < interfaces->count; i++) {
-		if (os_strcmp(interfaces->iface[i]->phy, phy) == 0) {
-			iface = interfaces->iface[i];
-			break;
-		}
-	}
-	if (iface && iface->state == HAPD_IFACE_DISABLED) {
-		for (i = 0; i < iface->num_bss; i++) {
-			h = iface->bss[i];
-			if (ifname[0] != '\0' &&
-			    !os_strncmp(ifname, h->conf->iface, sizeof(h->conf->iface)))
-				hapd_existed = true;
-		}
-	}
-	hostapd_config_free(conf);
-
-	for_each_mld_link(h, hapd)
-		old_valid_links |= BIT(h->mld_link_id);
-	hapd->mld->link_reconf_in_progress = old_valid_links;
-
-	if (hapd_existed)
-		ret = hostapd_enable_iface(iface);
-	else
-		ret = hostapd_add_iface(interfaces, cmd);
-	if (ret < 0)
-		goto out;
-
-	ret = os_snprintf(buf, buflen, "%s\n", "OK");
-	if (os_snprintf_error(buflen, ret))
-		ret = -1;
-	else
-		ret = 0;
-
-out:
-	os_free(tmp);
-
-	return ret;
-}
-
 static int hostapd_ctrl_iface_set_attlm(struct hostapd_data *hapd, char *cmd,
 					char *buf, size_t buflen)
 {
@@ -7238,8 +7143,7 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 						   reply, reply_size))
 			reply_len = -1;
 	} else if (os_strncmp(buf, "LINK_ADD ", 9) == 0) {
-		if (hostapd_ctrl_iface_link_add(hapd, buf + 9,
-						reply, reply_size))
+		if (hostapd_ctrl_iface_add(hapd->iface->interfaces, buf + 9) < 0)
 			reply_len = -1;
 	} else if (os_strncmp(buf, "SET_ATTLM ", 10) == 0) {
 		if (hostapd_ctrl_iface_set_attlm(hapd, buf + 10, reply,
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index b2c2a8852..1c757cb55 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -3674,7 +3674,7 @@ hostapd_interface_init_bss(struct hapd_interfaces *interfaces, const char *phy,
 			   const char *config_fname, int debug)
 {
 	struct hostapd_iface *new_iface = NULL, *iface = NULL;
-	struct hostapd_data *hapd;
+	struct hostapd_data *hapd, *tmp_hapd;
 	struct hostapd_config *conf;
 	int k;
 	size_t i, bss_idx;
@@ -3701,14 +3701,6 @@ hostapd_interface_init_bss(struct hapd_interfaces *interfaces, const char *phy,
 	if (!conf)
 		return NULL;
 
-#ifdef CONFIG_IEEE80211BE
-	/* AP MLD can be enabled with the same interface name, so even if we
-	 * get the interface, we still need to allocate a new hostapd_iface
-	 * structure. */
-	if (conf->bss[0]->mld_ap)
-		iface = NULL;
-#endif /* CONFIG_IEEE80211BE */
-
 	if (iface) {
 		struct hostapd_bss_config **tmp_conf;
 		struct hostapd_data **tmp_bss;
@@ -3723,17 +3715,23 @@ hostapd_interface_init_bss(struct hapd_interfaces *interfaces, const char *phy,
 		}
 
 		ifname = conf->bss[0]->iface;
-		if (conf->bss[0]->mld_ap) {
-			if (!iface->bss[0]->conf->mld_ap) {
-				wpa_printf(MSG_ERROR,
-					   "Cannot add a MLO BSS when the first BSS is non-MLO");
-				hostapd_config_free(conf);
-				return NULL;
-			}
-		} else {
-			if (ifname[0] != '\0' && ifname_in_use(interfaces, ifname)) {
+		if (ifname[0] == '\0') {
+			wpa_printf(MSG_ERROR,
+				   "Invalid interface name %s", ifname);
+			hostapd_config_free(conf);
+			return NULL;
+		}
+
+		tmp_hapd = hostapd_interfaces_get_hapd(interfaces, ifname);
+		if (tmp_hapd) {
+			/* A BSS with an existing interface name can only be added
+			 * into a different hostapd_iface as a AP MLD link
+			 */
+			if (!conf->bss[0]->mld_ap || !tmp_hapd->conf->mld_ap ||
+			    iface == tmp_hapd->iface) {
 				wpa_printf(MSG_ERROR,
-					   "Interface name %s already in use", ifname);
+					   "Interface name %s already in use",
+					   ifname);
 				hostapd_config_free(conf);
 				return NULL;
 			}
@@ -3798,6 +3796,28 @@ hostapd_interface_init_bss(struct hapd_interfaces *interfaces, const char *phy,
 		return NULL;
 	}
 
+#ifdef CONFIG_IEEE80211BE
+	for (i = bss_idx; i < iface->conf->num_bss; i++) {
+		tmp_hapd = iface->bss[i];
+
+		if (!tmp_hapd->conf->mld_ap)
+			continue;
+
+		/* If the newly added BSS belongs to an AP MLD and the AP MLD
+		 * is started, it's a link-add request.
+		 */
+		if (tmp_hapd->mld && tmp_hapd->mld->started) {
+			struct hostapd_data *link;
+			struct hostapd_mld *mld = tmp_hapd->mld;
+
+			mld->link_reconf_in_progress = 0;
+			for_each_mld_link(link, tmp_hapd)
+				mld->link_reconf_in_progress |= BIT(link->mld_link_id);
+
+		}
+	}
+#endif /* CONFIG_IEEE80211BE */
+
 	return iface;
 }
 
-- 
2.45.2

